packages/riverpod/lib/src/framework/base_provider.dart

```diff
@@ -608,6 +657,18 @@ class ProviderElement<Created, Listened> implements ProviderReference {
     }
   }
 
+  @override
+  void setState(Listened newState) {
+    if (!_mounted) {
+      throw StateError('Cannot call setState after a provider was dispose');
+    }
+    assert(_debugIsFlushing == false,
```

Comment:
debugIsFlushing is slightly different. I'd suggest using _debugCurrentlyBuildingElement for when "building", and creating a new debug flag for "onDispose" (inside _runOnDispose_)